File: crisis_app/lib/models/incident.dart
import 'location.dart';

class Incident {
  final String? id;
  final String type;
  final int severity;
  final Location location;
  final String source;
  final String status;
  final String? assignedTo;
  final String? timestamp;
  final bool escalated;
  final String? aiAnalysis;
  
  // NEW: Media Witness Fields
  final List<String>? evidenceImages;
  final String? evidenceAudio;
  final bool captureComplete;

  Incident({
    this.id,
    required this.type,
    required this.severity,
    required this.location,
    required this.source,
    required this.status,
    this.assignedTo,
    this.timestamp,
    this.escalated = false,
    this.aiAnalysis,
    this.evidenceImages,
    this.evidenceAudio,
    this.captureComplete = false,
  });

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      'type': type,
      'severity': severity,
      'location': location.toJson(),
      'source': source,
      'status': status,
      if (assignedTo != null) 'assigned_to': assignedTo,
      if (timestamp != null) 'timestamp': timestamp,
      'escalated': escalated,
      if (aiAnalysis != null) 'ai_analysis': aiAnalysis,
      if (evidenceImages != null) 'evidence_images': evidenceImages,
      if (evidenceAudio != null) 'evidence_audio': evidenceAudio,
      'capture_complete': captureComplete,
    };
  }

  factory Incident.fromJson(Map<String, dynamic> json) {
    final rawSeverity = json['severity'];
    final int parsedSeverity = rawSeverity is int
        ? rawSeverity
        : int.tryParse(rawSeverity?.toString() ?? '5') ?? 5;

    Location parsedLocation;
    try {
      final locRaw = json['location'];
      if (locRaw is Map) {
        parsedLocation = Location.fromJson(Map<String, dynamic>.from(locRaw));
      } else {
        parsedLocation = Location(lat: 0, long: 0, floor: 0);
      }
    } catch (_) {
      parsedLocation = Location(lat: 0, long: 0, floor: 0);
    }

    // Handle evidence_images if it's a dynamic list
    List<String>? images;
    if (json['evidence_images'] is List) {
      images = List<String>.from(json['evidence_images']);
    }

    return Incident(
      id: json['id']?.toString(),
      type: json['type']?.toString() ?? 'unknown',
      severity: parsedSeverity,
      location: parsedLocation,
      source: json['source']?.toString() ?? 'unknown',
      status: json['status']?.toString() ?? 'active',
      assignedTo: json['assigned_to']?.toString(),
      timestamp: json['timestamp']?.toString(),
      escalated: json['escalated'] == true,
      aiAnalysis: json['ai_analysis']?.toString(),
      evidenceImages: images,
      evidenceAudio: json['evidence_audio']?.toString(),
      captureComplete: json['capture_complete'] == true,
    );
  }
}

class SOSCreate {
  final String userId;
  final Location location;
  final String emergencyType;
  final dynamic phone;

  SOSCreate({
    required this.userId,
    required this.location,
    required this.emergencyType,
    this.phone,
  });

  Map<String, dynamic> toJson() {
    return {
      'user_id': userId,
      'location': location.toJson(),
      'emergency_type': emergencyType,
      if (phone != null) 'phone': phone,
    };
  }
}

File: backend/app/models/incident.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from enum import Enum

class IncidentType(str, Enum):
    FIRE = "fire"
    IOT = "iot"
    RADIO = "radio"
    SOS = "sos"
    PANIC = "panic"
    SMOKE = "smoke"

class IncidentSource(str, Enum):
    CCTV = "cctv"
    IOT = "iot"
    RADIO = "radio"
    APP = "app"
    GUEST = "guest"

class IncidentStatus(str, Enum):
    PENDING = "pending"
    ACTIVE = "active"
    ACKNOWLEDGED = "acknowledged"
    RESOLVED = "resolved"
    ESCALATED = "escalated"

class Location(BaseModel):
    lat: float
    long: float
    floor: Optional[int] = None

class IncidentBase(BaseModel):
    type: str 
    severity: int = Field(..., ge=1, le=10)
    location: Location
    source: str
    status: str = "active"
    escalated: bool = False
    assigned_to: Optional[str] = None
    ai_analysis: Optional[str] = None
    ai_error: Optional[str] = None
    triage_source: Optional[str] = None

    def toJson(self) -> Dict[str, Any]:
        return self.model_dump()

class IncidentCreate(IncidentBase):
    pass

class IncidentUpdate(BaseModel):
    type: Optional[str] = None
    severity: Optional[int] = None
    location: Optional[Location] = None
    status: Optional[str] = None
    escalated: Optional[bool] = None

class SOSCreate(BaseModel):
    user_id: str
    location: Location
    emergency_type: str
    source: Optional[str] = None
    phone: Optional[Any] = None

class IoTWebhookPayload(BaseModel):
    device_id: str
    event_type: str
    location: Location
    event_id: Optional[str] = None

class TriageInput(BaseModel):
    description: Optional[str] = None
    type: Optional[str] = "unknown"
    source: Optional[str] = "app"

class TriageResult(BaseModel):
    classification: str
    severity_score: int
    recommended_action: str
    action_plan: List[str] = []
    triage_source: str = "rule_based"
    ai_error: Optional[str] = None

class Incident(IncidentBase):
    id: Optional[str] = None
    timestamp: str

    @classmethod
    def from_dict(cls, data: Dict[str, Any]):
        return cls(**data)
